executor/docs/spec.md

```diff
@@ -0,0 +1,276 @@
+
+
+## Scheduler HTTP API
+
+The following enhancements to the existing Cook scheduler API will:
+
+1. Allow users to specify multiple commands when creating jobs
+2. Allow users to query the status code or codes of a completed job
+
+### Launch a Job
+
+It should be possible to specify multiple commands when launching a job. Any command failure will fail the whole job. In the future, it may be desirable to have finer-grained control over how failure will be handled (for example, ignoring failures from asynchronous monitoring commands).
+
+Request body schema additions:
+
+| JSON Key | Type | Description |
+|----------|------|-------------|
+| commands | array of command objects | the commands to run, as well as information about how they are to be run |
+
+`command` object schema:
+
+| JSON Key | Type | Description |
+|----------|------|-------------|
+| command  | string | the command to run |
+| async | boolean | a flag determining whether the command is run asynchronously.  defaults to `false` |
+| guard | boolean | a flag determining whether failure of the command will stop execution of the whole job. defaults to `false` |
+
+Example `commands`:
+
+
+    commands: [
+      {
+        async: true,
+        command: “tail_logs.py”
+      },
+      {
+        guard: true,
+        command: “ensure_only_one_job.sh”
+      },
+      {
+        command: "actual_command.py"
+      },
+      {
+        command: "send_email.pl"
+      }
+    ]
+
+Most return codes won’t change, but since `command` and `commands` are mutually exclusive, there’s a new way to get a `400` response.
+
+| Code | HTTP Meaning | Possible reason |
+|------|--------------|-----------------|
+| 400  | Malformed    | if both `command` and `commands` are set for a job |
+
+
+### Query Status of a Job
+
+Querying a job’s status also should return its exit code (or codes).
+
+Response schema additions:
+
+| JSON Key | Type | Description |
+|----------|------|-------------|
+| code | integer | status code when job specifies `command`  (will be absent if job specifies `commands`, or if job has yet to finish) |
+| codes | array of integers | status codes when job specifies `commands` (will be absent if job specifies `command`, or if job has yet to finish) |
+
+No changes to return codes are necessary.
+
+## Scheduler Datomic API
+
+It will also be necessary to change the Datomic schema in order to in order to support users of Cook’s “Datomic API”.
+
+Schema additions to support multiple commands per job:
+
+| Attribute | Type | Cardinality | Description |
+|-----------|------|-------------|-------------|
+| :job/command | ref | many | reference to a `command` entity |
+| :command/line | string | one | command line to run |
+| :command/code | double | one | exit code of command |
+| :command/params | ref | many | reference to a `command.param` entity |
+| :command.param/key | string | one | a command param key, like `async` |
```

Comment:
I would prefer to be explicit about these params (i.e. `:command/async`) since we are likely to only have a few of them and it is better from a datomic standpoint.